github/GithubRetry.py

```diff
@@ -0,0 +1,218 @@
+        """
+        self.secondaryRateWait = secondaryRateWait
+        # 403 is too broad to be retried, but GitHub API signals rate limits via 403
+        # we retry 403 and look into the response header via Retry.increment
+        # to determine if we really retry that 403
+        kwargs["status_forcelist"] = kwargs.get(
+            "status_forcelist", list(range(500, 600))
+        ) + [403]
+        kwargs["allowed_methods"] = kwargs.get(
+            "allowed_methods", Retry.DEFAULT_ALLOWED_METHODS.union({"GET", "POST"})
+        )
+        super().__init__(**kwargs)
+
+    def new(self, **kw):
+        kw.update(dict(secondaryRateWait=self.secondaryRateWait))
+        return super().new(**kw)
+
+    def increment(
+        self,
+        method=None,
+        url=None,
+        response=None,
+        error=None,
+        _pool=None,
+        _stacktrace=None,
+    ):
+        if response:
+            # we retry 403 only when there is a Retry-After header (indicating it is retry-able)
+            # or the body message does imply a rate limit error
+            if response.status == 403:
+                self.__log(
+                    logging.INFO,
+                    f"Request {method} {url} failed with {response.status}: {response.reason}",
+                )
+                if "Retry-After" in response.headers:
+                    # Sleeping 'Retry-After' seconds is implemented in urllib3.Retry.sleep() and called by urllib3
+                    self.__log(
+                        logging.INFO,
+                        f'Retrying after {response.headers.get("Retry-After")} seconds',
+                    )
+                else:
+                    content = response.reason
+
+                    # to identify retry-able methods, we inspect the response body
+                    try:
+                        content = self.get_content(response, url)
+                        content = json.loads(content)
+                        message = content.get("message")
+
+                        if Requester.isRateLimitError(message):
+                            rate_type = (
+                                "primary"
+                                if Requester.isPrimaryRateLimitError(message)
+                                else "secondary"
+                            )
+                            self.__log(
+                                logging.DEBUG,
+                                f"Response body indicates retry-able {rate_type} rate limit error: {message}",
+                            )
+
+                            # check early that we are retrying at all
+                            retry = super().increment(
+                                method, url, response, error, _pool, _stacktrace
+                            )
+
+                            # we backoff primary rate limit at least until X-RateLimit-Reset,
+                            # we backoff secondary rate limit at for secondaryRateWait seconds
```

Comment:
In my experience we can use `x-ratelimit-reset` for secondary wait as well, we don't have to guess